File: lab01/builder/src/main/java/com/example/builder/controller/CharacterController.java
package com.example.builder.controller;

import java.util.Map;

import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.example.builder.client.CharacterBuildClient;
import com.example.builder.dto.CharacterBuildRequest;
import com.example.builder.enums.CharacterType;
import com.example.builder.model.Character;

@RestController
@RequestMapping("/character")
public class CharacterController {

    private final CharacterBuildClient characterBuildClient = new CharacterBuildClient();

    @PostMapping("/with-builder")
    public Map<String, Object> createCharacterWithBuilder(@RequestBody CharacterBuildRequest request) {

        Character character = characterBuildClient.buildCharacter(request);

        return Map.of("response", character);
    }

    @PostMapping("/no-builder")
    public Map<String, Object> createCharacterWithoutBuilder(@RequestBody CharacterBuildRequest request) {
        Character character;

        if (request.getType() == CharacterType.ARCHER) {
            character = new Character(
                    request.getType(),
                    request.getName(),
                    request.getHealth(),
                    request.getMana(),
                    null
            );
        } else {
            character = new Character(
                    request.getType(),
                    request.getName(),
                    request.getHealth(),
                    request.getMana(),
                    request.getWeapon()
            );
        }

        return Map.of("response", character);
    }
}

File: lab01/builder1/Controllers/CharacterController.cs
using Builder1.Client;
using Builder1.Dto;
using Builder1.Enums;
using Builder1.Models;
using Microsoft.AspNetCore.Mvc;

namespace Builder1.Controllers;

[ApiController]
[Route("character")]
public class CharacterController : ControllerBase
{
    private readonly CharacterBuildClient _client = new();

    [HttpPost("with-builder")]
    public IActionResult CreateCharacterWithBuilder([FromBody] CharacterBuildRequest request)
    {
        Character character = _client.BuildCharacter(request);
        return Ok(new { response = character });
    }

    [HttpPost("no-builder")]
    public IActionResult CreateCharacterWithoutBuilder([FromBody] CharacterBuildRequest request)
    {
        Character character;

        if (request.Type == CharacterType.Archer)
        {
            character = new Character(request.Type, request.Name, request.Health, request.Mana, null);
        }
        else
        {
            character = new Character(request.Type, request.Name, request.Health, request.Mana, request.Weapon);
        }

        return Ok(new { response = character });
    }
}
